File: app/src/main/java/com/kkkkkn/readbooks/view/activities/BookInfoActivity.java
package com.kkkkkn.readbooks.view.activities;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.HeaderViewListAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.core.widget.ContentLoadingProgressBar;

import com.bumptech.glide.Glide;
import com.kkkkkn.readbooks.R;
import com.kkkkkn.readbooks.model.adapter.BookChaptersAdapter;
import com.kkkkkn.readbooks.model.adapter.SearchBookResultAdapter;
import com.kkkkkn.readbooks.model.entity.BookInfo;
import com.kkkkkn.readbooks.model.entity.ChapterInfo;
import com.kkkkkn.readbooks.presenter.Presenter_Info;
import com.kkkkkn.readbooks.presenter.Presenter_Login;
import com.kkkkkn.readbooks.util.ImageUtil;
import com.kkkkkn.readbooks.view.customView.CustomToast;
import com.kkkkkn.readbooks.view.view.BookInfoActivityView;


import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedList;


public class BookInfoActivity extends BaseActivity implements BookInfoActivityView {
    private final static String TAG="BookInfoActivity";
    private AppCompatTextView book_name,author_name,book_about;
    private AppCompatImageView book_img;
    private BookInfo bookInfo;
    private ArrayList<ChapterInfo> chapterList=new ArrayList<>();
    private ListView chapter_listView;
    private BookChaptersAdapter chaptersAdapter;
    private AppCompatButton btnStartRead,btnAddEnjoy;
    private boolean isEnd=false;
    private ContentLoadingProgressBar loading;
    private AppCompatTextView nothing_tv;


    private Presenter_Info presenter_info;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_book_info);

        initView();

        presenter_info=new Presenter_Info(getApplicationContext(),this);
        presenter_info.init();

        //查找图书信息是否存在
        bookInfo=(BookInfo)getIntent().getSerializableExtra("bookInfo");
        //获取图书信息
        if(bookInfo==null||bookInfo.isEmpty()){
            finish();
            return;
        }
        author_name.setText(bookInfo.getAuthorName());
        book_name.setText(bookInfo.getBookName());
        book_about.setText(bookInfo.getBookAbout());
        ImageUtil.loadImage(bookInfo.getBookImgUrl(),getApplicationContext(),book_img);
        //发送获取图书章节列表
        presenter_info.getBookChapters(chapterList.size(),bookInfo.getBookId());
    }

    //初始化绑定控件
    private void initView(){
        //绑定控件
        book_name=findViewById(R.id.bookInfo_bookName);
        author_name=findViewById(R.id.bookInfo_authorName);
        book_img=findViewById(R.id.bookInfo_bookImg);
        book_about=findViewById(R.id.bookInfo_bookAbout);
        chapter_listView=findViewById(R.id.bookInfo_chapters_listView);
        chaptersAdapter=new BookChaptersAdapter(chapterList,getApplicationContext());
        View footView;
        LayoutInflater inflater = LayoutInflater.from(this);
        footView = inflater.inflate(R.layout.loading_layout, null);
        loading = footView.findViewById(R.id.loading_view);
        nothing_tv = footView.findViewById(R.id.nothing_view);
        chapter_listView.addFooterView(footView);
        chapter_listView.setAdapter(chaptersAdapter);
        btnStartRead=findViewById(R.id.btn_StartRead);
        btnAddEnjoy=findViewById(R.id.btn_AddEnjoy);
        //跳转到浏览界面，从第一章开始阅读
        btnStartRead.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Bundle bundle=new Bundle();
                bundle.putSerializable("bookInfo",bookInfo);
                bundle.putSerializable("chapterInfo",chapterList.get(0));
                toBrowsingActivity(bundle);
            }
        });

        //添加到主页书架
        btnAddEnjoy.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //todo 弹窗或者动画效果 自定义加载框
                presenter_info.addBookShelf(bookInfo.getBookId());
            }
        });

        //章节点击跳转
        chapter_listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                Bundle bundle=new Bundle();
                bundle.putSerializable("chapterInfo",chapterList.get(i));
                bundle.putSerializable("bookInfo",bookInfo);
                toBrowsingActivity(bundle);
            }
        });

        chapter_listView.setOnScrollListener(new AbsListView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(AbsListView absListView, int i) {
                if (i == AbsListView.OnScrollListener.SCROLL_STATE_IDLE) {
                    // 判断是否滚动到底部
                    if (absListView.getLastVisiblePosition() == absListView.getCount() - 1) {
                        //是否还有未加载的数据
                        if(isEnd){
                            //listview 展示没有更多标签
                            nothing_tv.setVisibility(View.VISIBLE);
                        }else {
                            //显示加载框
                            nothing_tv.setVisibility(View.GONE);
                            loading.setVisibility(View.VISIBLE);
                            //加载更多功能的代码
                            presenter_info.getBookChapters(chapterList.size(),bookInfo.getBookId());
                        }
                    }
                }
            }

            @Override
            public void onScroll(AbsListView absListView, int i, int i1, int i2) { }
        });
    }

    @Override
    public void syncChapterList(final ArrayList<ChapterInfo> arrayList) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if (chapterList != null) {
                    if(arrayList.size()<presenter_info.getPageSize()){
                        isEnd=true;
                    }
                    loading.setVisibility(View.GONE);
                    chapterList.addAll(arrayList);
                    ((BookChaptersAdapter)((HeaderViewListAdapter)chapter_listView.getAdapter()).getWrappedAdapter()).notifyDataSetChanged();

                }
            }
        });
    }

    @Override
    public void toLoginActivity() {
        startActivity(new Intent(getApplicationContext(), LoginActivity.class));
    }

    @Override
    public void toBrowsingActivity(Bundle bundle) {
        Intent intent=new Intent(getApplicationContext(),BookBrowsingActivity.class);
        intent.putExtras(bundle);
        startActivity(intent);
    }

    @Override
    public void showMsgDialog(final int type,final String msg) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if(type>0){
                    CustomToast.showToast(getApplicationContext(),msg, Toast.LENGTH_SHORT,R.drawable.icon_msg_succese);
                }else {
                    CustomToast.showToast(getApplicationContext(),msg,Toast.LENGTH_SHORT,R.drawable.icon_msg_error);
                }
            }
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        chapterList.clear();
        presenter_info.release();
    }

}

File: app/src/main/java/com/kkkkkn/readbooks/view/activities/BookInfoActivity.kt
package com.kkkkkn.readbooks.view.activities

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import com.gyf.immersionbar.ImmersionBar
import com.kkkkkn.readbooks.databinding.ActivityBookInfoBinding
import com.kkkkkn.readbooks.model.adapter.BookChaptersAdapter
import com.kkkkkn.readbooks.model.entity.BookInfo
import com.kkkkkn.readbooks.model.entity.ChapterInfo
import com.kkkkkn.readbooks.presenter.PresenterInfo
import com.kkkkkn.readbooks.util.ImageUtil
import com.kkkkkn.readbooks.view.view.BookInfoActivityView
import com.scwang.smart.refresh.layout.api.RefreshLayout
import com.scwang.smart.refresh.layout.listener.OnLoadMoreListener
import es.dmoral.toasty.Toasty

class BookInfoActivity : BaseActivity<ActivityBookInfoBinding>(), BookInfoActivityView {
    private var bookInfo: BookInfo? = null
    private val chapterList: ArrayList<ChapterInfo> = ArrayList()
    private var chaptersAdapter: BookChaptersAdapter? = null
    private var presenterInfo: PresenterInfo? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initView()
        presenterInfo = PresenterInfo(applicationContext, this)
        presenterInfo!!.init()

        //查找图书信息是否存在
        bookInfo = intent.getSerializableExtra("bookInfo") as BookInfo?
        //获取图书信息
        if (bookInfo == null || bookInfo!!.isEmpty) {
            finish()
            return
        }
        mViewBinding.bookInfoAuthorName.text = bookInfo!!.authorName
        mViewBinding.bookInfoBookName.text = bookInfo!!.bookName
        mViewBinding.bookInfoBookAbout.text = bookInfo!!.bookAbout
        bookInfo!!.bookImgUrl?.let {
            ImageUtil.loadImage(
                it,
                applicationContext, mViewBinding.bookInfoBookImg
            )
        }
        //发送获取图书章节列表
        presenterInfo!!.getBookChapters(chapterList.size, bookInfo!!.bookId)
    }

    //初始化绑定控件
    private fun initView() {
        //关闭下拉刷新
        mViewBinding.infoChapterRefresh.setEnableRefresh(false)
        mViewBinding.infoChapterRefresh.setOnLoadMoreListener(object :OnLoadMoreListener{

            override fun onLoadMore(refreshLayout: RefreshLayout) {
                //发送获取图书章节列表
                presenterInfo!!.getBookChapters(chapterList.size, bookInfo!!.bookId)
            }

        })
        
        //绑定控件
        chaptersAdapter = BookChaptersAdapter(chapterList, this)
        mViewBinding.bookInfoChaptersListView.layoutManager= LinearLayoutManager(this)
        mViewBinding.bookInfoChaptersListView.adapter = chaptersAdapter
        //章节点击跳转
        chaptersAdapter!!.setItemOnClickListener(object :BookChaptersAdapter.ItemOnClickListener{
            override fun onItemClick(position: Int) {
                val bundle = Bundle()
                bundle.putSerializable("chapterInfo", chapterList[position])
                bundle.putSerializable("bookInfo", bookInfo)
                Log.i(tag, "initView: 开始跳转 chapterInfo "+chapterList[position].chapterPath+" bookInfo "+bookInfo.toString())
                toBrowsingActivity(bundle)
            }

        })

        //跳转到浏览界面，从第一章开始阅读
        mViewBinding.btnStartRead.setOnClickListener {
            val bundle = Bundle()
            bundle.putSerializable("bookInfo", bookInfo)
            bundle.putSerializable("chapterInfo", chapterList[0])
            toBrowsingActivity(bundle)
        }

        //添加到主页书架
        mViewBinding.btnAddEnjoy.setOnClickListener { //todo 弹窗或者动画效果 自定义加载框
            presenterInfo!!.addBookShelf(bookInfo!!.bookId)
        }


    }

    override fun syncChapterList(linkedList: ArrayList<ChapterInfo>) {
        runOnUiThread {
            if(mViewBinding.infoChapterRefresh.isLoading){
                //判断已无更多数据加载
                if (linkedList.size < presenterInfo!!.pageSize) {
                    mViewBinding.infoChapterRefresh.finishLoadMoreWithNoMoreData()
                }else{
                    mViewBinding.infoChapterRefresh.finishLoadMore(true)
                }
            }
            val start=chapterList.size
            chapterList.addAll(linkedList)
            mViewBinding.bookInfoChaptersListView.adapter?.notifyItemRangeChanged(start,chapterList.size)
        }
    }

    override fun toLoginActivity() {
        startActivity(Intent(applicationContext, LoginActivity::class.java))
    }

    override fun toBrowsingActivity(bundle: Bundle) {
        val intent = Intent(applicationContext, BookBrowsingActivity::class.java)
        intent.putExtras(bundle)
        startActivity(intent)
    }

    override fun showMsgDialog(type: Int, msg: String) {
        runOnUiThread {
            if (type > 0) {
                Toasty.success(applicationContext, msg, Toast.LENGTH_SHORT, true).show()
            } else {
                Toasty.error(applicationContext, msg, Toast.LENGTH_SHORT, true).show()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        chapterList.clear()
        if (presenterInfo != null) {
            presenterInfo!!.release()
        }
    }

    override fun getViewBinding(): ActivityBookInfoBinding {
        return ActivityBookInfoBinding.inflate(layoutInflater)
    }

    override fun editStatusBar(){
        ImmersionBar.with(this)
            .statusBarDarkFont(true)
            .navigationBarDarkIcon(true)
            .init()
    }
}
